flink-runtime/src/test/java/org/apache/flink/runtime/rest/RestClientTest.java

```diff
@@ -89,7 +91,78 @@ public void testInvalidVersionRejection() throws Exception {
 		} catch (IllegalArgumentException e) {
 			// expected
 		}
+	}
 
+	/**
+	 * Tests that we fail the operation if the remote connection closes.
+	 */
+	@Test
+	public void testConnectionClosedHandling() throws Exception {
+		final ServerSocket serverSocket = new ServerSocket(0);
+		final String targetAddress = "localhost";
+		final int targetPort = serverSocket.getLocalPort();
+
+		try (final RestClient restClient = new RestClient(RestClientConfiguration.fromConfiguration(new Configuration()), TestingUtils.defaultExecutor())) {
+			final CompletableFuture<EmptyResponseBody> responseFuture = restClient.sendRequest(
+				targetAddress,
+				targetPort,
+				new TestMessageHeaders(),
+				EmptyMessageParameters.getInstance(),
+				EmptyRequestBody.getInstance(),
+				Collections.emptyList());
+
+			// establish connection
+			final Socket connectionSocket = serverSocket.accept();
+
+			// close connection
+			connectionSocket.close();
+
+			try {
+				responseFuture.get();
+			} catch (ExecutionException ee) {
+				assertThat(ExceptionUtils.findThrowable(ee, ConnectionClosedException.class).isPresent(), is(true));
+			}
+		} finally {
+			serverSocket.close();
+		}
+	}
+
+	/**
+	 * Tests that we fail the operation if the client closes.
+	 */
+	@Test
+	public void testRestClientClosedHandling() throws Exception {
+		final ServerSocket serverSocket = new ServerSocket(0);
+		final String targetAddress = "localhost";
+		final int targetPort = serverSocket.getLocalPort();
+		Socket connectionSocket = null;
+
+		try (final RestClient restClient = new RestClient(RestClientConfiguration.fromConfiguration(new Configuration()), TestingUtils.defaultExecutor())) {
+			final CompletableFuture<EmptyResponseBody> responseFuture = restClient.sendRequest(
+				targetAddress,
+				targetPort,
+				new TestMessageHeaders(),
+				EmptyMessageParameters.getInstance(),
+				EmptyRequestBody.getInstance(),
+				Collections.emptyList());
+
+			// establish connection
+			connectionSocket = serverSocket.accept();
+
+			restClient.close();
+
+			try {
+				responseFuture.get();
+			} catch (ExecutionException ee) {
+				assertThat(ExceptionUtils.findThrowable(ee, ConnectionClosedException.class).isPresent(), is(true));
```

Comment:
this test is currently failing, and the travis logs aren't that helpful.
We could write the assertion like this to make the error message more helpful:
```
optionalException = ExceptionUtils.findThrowable(ee, ConnectionClosedException.class)
if (!optionalException.isPresent()) {
	throw ee;
}
```

Reply:
yes, there were actually a bunch of related problems: Channel inactivity and not initialized pipeline.